initial take on ADR reasoning the move from Azure managed db to in-cl… (#583)
danskernesdigitalebibliotek/dpl-platform: docs/architecture/adr/adr-006-in-cluster-db.md

managed by Azure. Azure is barring us access to the database server logs due
to their disproportionate pricing of server logs. This results in an inability
to investigate database slowness as well as what lead to a database crash.
Instead we've had to wait for Azure support to allocate time and resources
towards advising/assisting us.

We have also experienced the Azure support taking unsanctioned action on the
database, resulting in downtime for all sites, on serveral occasions.

The sum of all this is an erroded trust in Azure ability to cater adequately
to our needs.

By moving the database into the cluster, we gain access to the database server
logs, access to the server itself and a guarantee that no one but the platform
team takes actions and thus.
We have the experience running databases in cluster, and having the power and
agency to do the things we know need doing, should result in less down time for
the sites as well as the ability to bring the sites backup faster.

### Pros & Cons

#### Azure Managed DB

##### Cons

- Experienced to be poorly managed
- Server logs is disproportionately priced = too expensive for this project
- We don't have access to the server

> - The wait time for support can be long and very bothersome as well as

Reviewer: "and the support given can be incorrect". It's not the wait time that's incorrect.

Author: fixed  - good change

incorrect.
- Server might be (was in our case) misconfigured
- Will require a migration to a potentially as bad service

##### Pros

- One-click setup
- Minimal configuration needed from us
- Azure support
- Azure can be blamed for downtime, they are directly or indirectly responsible
for.

> - MySql, which is not set to be sunsettet, might have a noticably better service

Reviewer: "sunset"

Author: fixed

# ADR 006: In-cluster Database

## Context

> Microsoft has decided to sunset Azure Database for MariaDB summer 2025.

Reviewer 2: `has decided to sunset` -> `is sunsetting` (simplify)

> They recommend moving to their Azure Database for MySql instead.

Reviewer 2: Remove "their" (simplify)

The project has experienced project wide down time due to database crashes on
multiple occasions:

- 17.01.2025: Database went down due to a restart initiated by the Azure Product
Group team taking unsanctioned action
- 15.01.2025: Database went down due to a restart initiated by the Azure Product
Group team taking unsanctioned action
- 9.01.2025: Database went down as a result of to many login requests
- 28.11.2024: Database went down due to too many active connection attempts
- 7.11.2024: Database went down due to too many connections
- Summer 2024: Database was restarted by Azure for unscheduled and unnotified
maintenance

> Microsofts datbase server logs are prohibitively expensive.

Reviewer 2: `Microsoft{->'}s dat{->a}base` (typos)

Microsoft misconfigured the database provision-time.
Microsofts support is tiered and direct access to technical staff is prevented

> by Azure support policy.

Reviewer 2: I think the important point here is that the current support tier is useless, and useful support is prohibitively expensive. It is unclear what the point is from the current sentence.

> Microsofts support has made unsanctioned changes on the databse

Reviewer 2: `Micorsoft{s->} support` (typo)

Reviewer 2: `datab{->a}se`

Maybe exemplify explicitly what they did despite being told not to.

> It is not possible to access the database server.

Reviewer 2: What does this mean?

Author: We don't have access... we can go and do stuff on the server

Reviewer 3: I do not think this a reasonable expectation to have for a managed database product.

Author: refactored

# ADR 006: In-cluster Database

## Context

> Microsoft is sunsetting Azure Database for MariaDB summer 2025.

Author: added

Group team taking unsanctioned action
- 15.01.2025: Database went down due to a restart initiated by the Azure Product
Group team taking unsanctioned action
- 9.01.2025: Database went down as a result of to many login requests
- 28.11.2024: Database went down due to too many active connection attempts
- 7.11.2024: Database went down due to too many connections
- Summer 2024: Database was restarted by Azure for unscheduled and unnotified
maintenance

Microsoft's database server logs are prohibitively expensive.
Microsoft misconfigured the database provision-time.
The current Microsoft support tier is useless and useful support is
prohibitively expensive.
Microsoft support has made unsanctioned changes on the databse, such as
maintenance work during business hours and restarts in attempts to fix support
tickets.
It is not possible to access the database server.

Installing a MariaDB Sql operator in the cluster provides the following
benefits:

- access to the database server logs
- access to the server itself
- a guarantee that no one but the platform team can take action on the server
- ability to right-size the database
- Running clusters of databases
- Database recovery

> - The ability to start recovering from a crash immediately

Reviewer 3: Also: I think it is worth adding a bit more context into what we mean by "In-cluster database".

I am not an expert here but as I see it there might be multiple ways to run a database in a cluster. I think it would be good if we could provide more info regarding what we actually have in mind, preferably combined with a link or two if we are referring to specific solutions/products/architectures e.g. [Kubernetes Operator for MariaDB](https://mariadb.com/kb/en/kubernetes-operators-for-mariadb/).

> - The ability to start recovering from a crash immediately

Author: > The database is a critical part of the sites managed by the platform. Downtime or degradation of service for the database is usually reflected to all applications running on the platform.
This is already in the document. Do you find it unclear? 

> Can we provide any reasoning for why we do not seriously consider other alternatives? 
I don't think I know of any architecturally sound alternatives to the two options we have at hand. 

> Also: I think it is worth adding a bit more context into what we mean by "In-cluster database".
Please elaborate on what you think should be mentioned

>  I think it would be good if we could provide more info regarding what we actually have in mind,
I agree, but we haven't decided on an architecture yet. We should circle back to this when we have made a decision.

### Pros & Cons

#### Azure Managed DB

##### Cons

- Experienced to be poorly managed
- Server logs is disproportionately priced = too expensive for this project
- We don't have access to the server

> - The wait time for support can be long and incorrect.

Author: In part this is an attempt at shorten the experience with Azure Support. I do believe on of the issues with the Support is that we have to wait for them to have time to help us. In contrast, having the databases in-cluster, will allow immediate troubleshooting

> - The wait time for support can be long and incorrect.

Author: Is it the first list going over the bad experiences we've had?

> - Server might be (was in our case) misconfigured

Author: The server was configured to match that of much more powerful server than the one we have, so I would say they definitely misconfigured our server.

> - Server might be (was in our case) misconfigured

Author: Is it the first list going over the bad experiences we've had?

> - Will require a migration to a potentially as bad service

Author: refactored

##### Pros

- One-click setup
- Minimal configuration needed from us
- Azure support
- Azure can be blamed for downtime, they are directly or indirectly responsible
for.
- MySql, which is not set to be sunset, might have a noticably better service

> as that is their chosen flavor to keep offering.

Reviewer 3: I think all this can be reduced to something like "Continual use of managed database product".

Author: all of the bullet points?

> as that is their chosen flavor to keep offering.

Reviewer 3: I suggest adding that Azure Database for MySQL supports high availability with automatic failover to a replica as a pro

Author: done

#### In-Cluster DB

##### Cons

- We have no one but ourselves to blame if something goes awry

> - Logging and Monitoring has to be set up manually

Reviewer 3: ```suggestion
- Logging and monitoring has to be set up manually
```

- All configuration has to be done ourselves

> - Not possible to do PITR recovery at the moment.

Reviewer 3: ```suggestion
- [Point-in-time-recovery is not supported at the moment](https://github.com/mariadb-operator/mariadb-operator/issues/507)
```

> - Not possible to do PITR recovery at the moment.

Reviewer 3: Consider adding a point that we have to manage software update/security/whatever ourselves. In other words: It is our own responsibility to keep the version of the database up to date and react to security updates in a timely fashion.

Author: done

##### Pros

- Ability to investigate logs
- Logging and monitoring can be set up to alert us to noteworthy changes

> - A little cheaper on the face of it

Reviewer 3: Do we really want to highlight this as a pro? 

I do not think our current spending on MariaDB is that high. Establishing this new solution will result in additional costs for setup and maintenance.

Author: The rate is cheaper, and more database power will therefore cost less than Azure database for MariaDB og MySql. I rephrase it to highlight this

- Ability to split out databases such that one database having issues doesn't

> cause all sites to crash.

Reviewer 3: :+1: I think this is an important pro. I do not know if the current structure is a result of the [Lagoon DBaaS controller](https://github.com/uselagoon/dbaas-controller) but it is a nice change in regard to removing a single point of failure.

Author: I agree - it will be huge plus

> - Possible to run in a HA setup if we should decide to do so.

Reviewer 3: ```suggestion
- Support for high-availability if we decide to implement this.
```

Can we provide link(s) to what approach we have in mind here?

- Performance can be optimized and finetuned to our usecase
- Databases will be located closer, logically as well as physically, to
workloads relying on them = faster response time, which should be noticable for

> end-users.

Reviewer 3: :+1:

- We can right-size the databae, thereby getting the maximum performance for the

> buck.

Reviewer 3: As mentioned previously: Personally I do not see money as a key driver here. The database is far from the most expensive service we use.

Reviewer 2: I agree that it isn't the most important driver, but I think it is worth considering along with all the other factors. If running our own database was more expensive, it would definitely be a factor.

## Decision

We have made the decision to implement the In-cluster database due the
afforemention pros and cons lists.

> We are tired of having a poorly managed database, with bad support.
> Additionally both, we and the DDF, as well as the libraries themselves are
> tired of having preventable downtime.
> This way we take our agency back and ensure that we will not have an incoming
> task of migrating the database later in the year, when Azure End-Of-Lifes
> Managed MariaDB's.

Reviewer 3: ```suggestion
Negative experiences with stability and management for Azure Database for MariaDB have reduced our trust in continual use of managed databases by Azure.
By moving to an in-cluster database we take our agency for this critical part of the infrastructure back.
```

We will start by implementing a PoC, where we can test for a good setup of the in-cluster
database before migrating all the databases to use the in-cluster databases.
The PoC will also be used for testing mgiration against.

Every step taken towards moving into in-cluster database shall done
transparantly.

The In-cluster database must:

> - have ressource monitoring setup
> - be able to have backups taken

Reviewer 3: ```suggestion
- Have ressource monitoring setup
- Be able to have backups taken
```

Casing for consistency.

# ADR 006: In-cluster Database

## Context

We have known, since summer 2024, that Microsoft had decided to sunset Azure
database for MariaDB. Their recommendation has been to migrate to their Flexible

> MySql database which they have provided pretty well documented. Back then we had

Reviewer 2: "which they have provided pretty well documented" - I don't know what this means, but it can also just be removed. The rest of the sentence is good and fine.

already experienced database downtime, which prompted us to start a discussion

> about wether to move the databse to Azure Flexible MySql server or to an

Reviewer 2: `w{->h}ether`

in-cluster database. This conversation has been running for the better part of

> 6-8 months.

Reviewer 2: I'm not sure all the context of how long we have been discussing this is relevant to the final decision. I would just jump straight into the points.

The main points as to why we are deciding not to follow Azure's recommendation, as I see it, is the stability issues with MariaDB and poor support when things went wrong.

I think the intro would be clearer if it had just these points in order (with your written out wording, of course):

- MariaDB is being sunsetted
- Microsoft recommends MySQL flexible
- Microsoft's operation of MariaDB was very unstable
- Microsoft's support when things went wrong was bad
- So we decided to evaluate alternatives

The main points we kept debating was:

- Can we operate an in-cluster database on par or better than Azure?
- Are we able to deal with problems that might arise on the databases with out
Azures Technical Support?
- If we self host the databse, are we then able to optimize the performance?
- Can we guarantee a better or on par uptime to Azure's managed database if we
selfhost the database?
- Would we get better support if we moved to the suggested managed database?
- Would the performance be better on the database suggested by Azure?

> - Would downtime be minimized if we moved to the sugggested manged database?

Reviewer 2: I don't think a list of question works well in a decision document. Stick to the answers instead of the questions.

Maybe these questions can be turned into headings for sections you go through to discuss the issue?

We have continually, and to a worsening degree, seen even more downtime since
summer. We have been cut-off from having insight into the logs of the database
due to an extremely steep price on server logs, which we tried enabling during
the summer. This resulted in a more than doubled operations cost, which the

> operations budget in now way can bear.

Reviewer 2: I think this comes across as subjective and imprecise.

Is it important to talk about all the issues with MariaDB, and how bad it was? I think all of the issues could be summarized more briefly.

Then, I think it is more important to talk about how these issues shape our view of and expectations for Microsoft-run services in general.

many connections" and "Ressource busy", which should have resulted in end-user
experiences such as 500 errors and not project/nation wide downtime.

Conversations with the Azure Technical Support is seeped in effective solution
conversations. This is due to:

- Language barriers as the Azure support is handled by people whos English
proficiency is wildly varying.
- The tiered support which bars us from talking to the actual technical staff.
This results in a knowledge relay bottleneck, often resulting in
lost-in-translation situations.
- Tranparency issues stemming from somewhere inside of the technical support,
where information is withheld in such a way that it builds mistrust to Azure.

Lastly we have experienced Azure Technical Support taking unsanctioned actions
on the database, resulting in project wide downtime on serveral occasions. The
first time we exprienced this was summer 2024, where they restarted the databse
while doing some configuration. That last time was january 2025, where they
restarted the database multiple times trying to fix a bad database recovery.

Deranged has dealt with managed databases in the past, and has never experienced

> a level of service as bad as this.

Reviewer 2: Too much. Too subjective. Tone should be more objective and sober.

yes, we're frustrated - but that is irrelevant to the decision.

Deranged has been managing databases in-cluster for the last 7 years. During
this time, Deranged has managed the following database types:

- Postgress
- MongoDB
- MariaDB/MySql
- ElasticSearch
- Redis

> - InfluxDB

Reviewer 2: This isn't relevant to the decision.
